File: src/main/java/at/yawk/math/data/Expressions.java
package at.yawk.math.data;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;
import lombok.experimental.UtilityClass;

/**
 * @author yawkat
 */
@UtilityClass
public class Expressions {
    public static Expression add(Expression lhs, Expression rhs) {
        return new AdditionExpression(lhs, rhs);
    }

    public static Expression multiply(Expression lhs, Expression rhs) {
        return new MultiplicationExpression(lhs, rhs);
    }

    public static Expression subtract(Expression lhs, Expression rhs) {
        return add(lhs, negate(rhs));
    }

    public static Expression divide(Expression lhs, Expression rhs) {
        return multiply(lhs, reciprocal(rhs));
    }

    public static Expression negate(Expression expression) {
        return multiply(integer(-1), expression);
    }

    public static Expression reciprocal(Expression expression) {
        return new ReciprocalExpression(expression);
    }

    public static Vector vector(List<Expression> expressions) {
        return Vector.builder()
                .expressions(expressions)
                .build();
    }

    public static Vector vector(Expression... expressions) {
        return vector(Arrays.asList(expressions));
    }

    public static IntegerExpression integer(long value) {
        return new IntegerExpression(value);
    }

    public static IntegerExpression integer(BigInteger value) {
        return new IntegerExpression(value);
    }

    public static Expression rational(long numerator, long denominator) {
        return divide(integer(numerator), integer(denominator));
    }

    public static Expression abs(Expression expression) {
        return new AbsoluteValueExpression(expression);
    }
}

File: src/main/kotlin/at/yawk/math/data/Expressions.kt
package at.yawk.math.data

import java.math.BigInteger

/**
 * @author yawkat
 */
object Expressions {
    val zero = int(0)
    val one = int(1)
    val minusOne = int(-1)

    fun add(vararg a: Expression): Expression {
        return AdditionExpression(listOf(*a))
    }

    fun multiply(vararg a: Expression): Expression {
        return MultiplicationExpression(listOf(*a))
    }

    fun subtract(minuend: Expression, subtrahend: Expression): Expression {
        return add(minuend, negate(subtrahend))
    }

    fun divide(dividend: Expression, divisor: Expression): Expression {
        return multiply(dividend, reciprocal(divisor))
    }

    fun negate(a: Expression): Expression {
        if (a is RealNumberExpression) return a.negate
        return multiply(minusOne, a)
    }

    fun reciprocal(a: Expression): Expression {
        if (a is RealNumberExpression) return a.reciprocal
        return ExponentiationExpression(a, minusOne)
    }

    fun vector(rows: List<Expression>): Expression {
        return Vector(rows)
    }

    fun int(value: BigInteger): IntegerExpression {
        return IntegerExpression(value)
    }

    fun int(value: Long): IntegerExpression {
        return int(BigInteger.valueOf(value))
    }

    fun rational(numerator: Long, denominator: Long): Rational {
        return SimpleRational(int(numerator), int(denominator))
    }

    fun rational(numerator: BigInteger, denominator: BigInteger): Rational {
        return SimpleRational(int(numerator), int(denominator))
    }

    fun abs(expression: Expression): Expression {
        return AbsoluteValueExpression(expression)
    }

    fun gcd(a: Expression, b: Expression): Expression {
        return GcdExpression(a, b)
    }

    fun lcm(a: Expression, b: Expression): Expression {
        return LcmExpression(a, b)
    }

    fun dotProduct(a: Expression, b: Expression): Expression {
        return DotProductExpression(a, b)
    }

    fun pow(base: Expression, exponent: Expression): Expression {
        return ExponentiationExpression(base, exponent)
    }
}